File: CoreLib/ImplPAL/Win32/Win32Native.cs
namespace Microsoft.Win32
{
    using System;
    using System.Security;
    using System.Text;
    using System.Threading;
    using Microsoft.Win32.SafeHandles;
    using System.Runtime.CompilerServices;
    using System.Runtime.ConstrainedExecution;

    using BOOL = System.Int32;
    using DWORD = System.UInt32;
    using ULONG = System.UInt32;

    /**
     * Win32 encapsulation for MSCORLIB.
     */
    // Remove the default demands for all P/Invoke methods with this
    // global declaration on the class.

    [SuppressUnmanagedCodeSecurityAttribute()]
    internal static partial class Win32Native
    {
        private const int STDIN_FILENO = 0;

        private const int STDOUT_FILENO = 1;

        private const int STDERR_FILENO = 2;

        private const int O_RDONLY = 0x0000;	/* open for reading only */

        private const int O_WRONLY = 0x0001;	/* open for writing only */

        private const int O_RDWR = 0x0002;	/* open for reading and writing */

        private const int O_CREAT = 0x0100;		/* create if nonexistant */

        private const int O_TRUNC = 0x0200;		/* truncate to zero length */

        private const int O_EXCL = 0x0040;		/* error if already exists */

        private const int F_SETFD = 2;		/* set file descriptor flags */

        private const int LOCK_SH = 0x01;		/* shared file lock */

        private const int LOCK_EX = 0x02;		/* exclusive file lock */

        private const int LOCK_NB = 0x04;		/* don't block when locking */

        private const int LOCK_UN = 0x08;		/* unlock file */

        /* access function */
        private const int F_OK = 0;	/* test for existence of file */

        private const int O_DIRECT = 00040000;

        private const int O_BINARY = 0x8000;

        private const int S_IRUSR = 0000400;			/* R for owner */

        private const int S_IWUSR = 0000200;			/* W for owner */

        private const int S_IROTH = 0000004;			/* R for other */

        private const int S_IRGRP = 0000040;			/* R for group */

        private const uint GENERIC_READ = 0x80000000;

        private const uint GENERIC_WRITE = 0x40000000;

        public const uint FILE_ATTRIBUTE_HIDDEN = 0x00000002;

        public const uint FILE_ATTRIBUTE_SYSTEM = 0x00000004;

        //public const uint FILE_ATTRIBUTE_DIRECTORY = 0x00000010;

        public const uint FILE_ATTRIBUTE_ARCHIVE = 0x00000020;

        public const uint FILE_ATTRIBUTE_DEVICE = 0x00000040;

        public const uint FILE_ATTRIBUTE_NORMAL = 0x00000080;

        private const uint S_IFMT = 0xF000;

        private const uint S_IFDIR = 0x4000;

        private const int FILE_FLAG_NO_BUFFERING = 0x20000000;

        [MethodImpl(MethodImplOptions.Unmanaged)]
        public unsafe static extern byte* __get_full_path(byte* file_name, byte* resolved_name);

        [MethodImplAttribute(MethodImplOptions.Unmanaged)]
        private extern unsafe static int open(byte* fileName, int flags, int mode);

        [MethodImplAttribute(MethodImplOptions.Unmanaged)]
        private extern static int close(int fd);

        [MethodImpl(MethodImplOptions.Unmanaged)]
        public unsafe static extern int write(int fd, void* buf, int count);

        [MethodImpl(MethodImplOptions.Unmanaged)]
        public unsafe static extern int read(int fd, void* buf, int count);

        [MethodImpl(MethodImplOptions.Unmanaged)]
        public static extern int fcntl(int fd, int cmd, __arglist);

        [MethodImpl(MethodImplOptions.Unmanaged)]
        public static extern int flock(int fd, int operation);

        [MethodImpl(MethodImplOptions.Unmanaged)]
        public static unsafe extern int access(byte* pathname, int mode);

        [MethodImpl(MethodImplOptions.Unmanaged)]
        public static unsafe extern int stat(byte* path, int* buf);

        [MethodImpl(MethodImplOptions.Unmanaged)]
        public static unsafe extern int fstat(int fd, int* buf);

        private static int _errorMode;

        struct stat_data
        {
            public int st_dev;     /* ID of device containing file */
            public int st_ino;     /* inode number */
            public ushort st_mode;    /* protection */
            public short st_nlink;   /* number of hard links */
            public short st_uid;     /* user ID of owner */
            public short st_gid;     /* group ID of owner */
            public int st_rdev;    /* device ID (if special file) */
            public int st_size;    /* total size, in bytes */
            public int st_atime;   /* time of last access */
            public int st_mtime;   /* time of last modification */
            public int st_ctime;   /* time of last status change */
            public int reserved0;
            public int reserved1;
            public int reserved2;
            public int reserved3;
            public int reserved4;
            public int reserved5;
            public int reserved6;
        };

        public static byte[] ToAsciiString(string s)
        {
            if (s == null)
            {
                throw new ArgumentNullException("s", "String");
            }

            var byteCount = Encoding.ASCII.GetByteCount(s);
            // +1 needed for ending \0
            var bytes = new byte[byteCount + 1];
            var bytesReceived = Encoding.ASCII.GetBytes(s, 0, s.Length, bytes, 0);
            return bytes;
        }

        public static byte[] ToAsciiString(char[] chars)
        {
            if (chars == null)
            {
                throw new ArgumentNullException("chars", "Chars");
            }

            var byteCount = Encoding.ASCII.GetByteCount(chars);
            // +1 needed for ending \0
            var bytes = new byte[byteCount + 1];
            var bytesReceived = Encoding.ASCII.GetBytes(chars, 0, chars.Length, bytes, 0);
            return bytes;
        }

        internal static bool SetEvent(SafeWaitHandle handle)
        {
            var acquiredLock = false;

            try
            {
                Monitor.Enter(handle, ref acquiredLock);

                unsafe
                {
                    var waitHandleData = (int*)handle.DangerousGetHandle().ToPointer();
                    if (waitHandleData == null)
                    {
                        throw new InvalidOperationException();
                    }

                    // state = true
                    waitHandleData[0] = 1;

                    // type = AutoReset
                    if (waitHandleData[1] > 0)
                    {
                        Monitor.Pulse(handle);
                    }
                    else
                    {
                        Monitor.PulseAll(handle);
                    }
                }
            }
            finally
            {
                if (acquiredLock)
                {
                    Monitor.Exit(handle);
                }
            }

            return true;
        }

        internal static bool ResetEvent(SafeWaitHandle handle)
        {
            var acquiredLock = false;

            try
            {
                Monitor.Enter(handle, ref acquiredLock);

                unsafe
                {
                    var waitHandleData = (int*)handle.DangerousGetHandle().ToPointer();
                    if (waitHandleData == null)
                    {
                        throw new InvalidOperationException();
                    }

                    // state = false
                    waitHandleData[0] = 0;
                }
            }
            finally
            {
                if (acquiredLock)
                {
                    Monitor.Exit(handle);
                }
            }

            return true;
        }

        internal static SafeWaitHandle CreateEvent(SECURITY_ATTRIBUTES lpSecurityAttributes, bool isManualReset, bool initialState, String name)
        {
            // state, type
            var waitHandleData = new int[2];

            waitHandleData[0] = initialState ? 1 : 0;
            waitHandleData[1] = isManualReset ? 0 : 1;

            // we use this SafeWaitHandle as SyncBlock object
            unsafe
            {
                fixed (int* ptr = waitHandleData)
                {
                    return new SafeWaitHandle(new IntPtr(ptr), true);
                }
            }
        }

        private static unsafe int wcslen(char* ptr)
        {
            char* end = ptr;

            // The following code is (somewhat surprisingly!) significantly faster than a naive loop,
            // at least on x86 and the current jit.

            // First make sure our pointer is aligned on a dword boundary
            while (((uint)end & 3) != 0 && *end != 0)
                end++;
            if (*end != 0)
            {
                // The loop condition below works because if "end[0] & end[1]" is non-zero, that means
                // neither operand can have been zero. If is zero, we have to look at the operands individually,
                // but we hope this going to fairly rare.

                // In general, it would be incorrect to access end[1] if we haven't made sure
                // end[0] is non-zero. However, we know the ptr has been aligned by the loop above
                // so end[0] and end[1] must be in the same page, so they're either both accessible, or both not.

                while ((end[0] & end[1]) != 0 || (end[0] != 0 && end[1] != 0))
                {
                    end += 2;
                }
            }
            // finish up with the naive loop
            for (; *end != 0; end++)
                ;

            int count = (int)(end - ptr);

            return count;
        }

        internal unsafe static int GetFullPathName(char* path, int numBufferChars, char* buffer, IntPtr mustBeZero)
        {
            if (path == null)
            {
                throw new ArgumentNullException("path", "path");
            }

            var byteCount = Encoding.ASCII.GetByteCount(path, wcslen(path));
            // +1 needed for ending \0
            var relative_path_ascii = stackalloc byte[byteCount + 1];

            var bytesReceived = Encoding.ASCII.GetBytes(path, wcslen(path), relative_path_ascii, byteCount);

            var resolved_path_ascii = stackalloc byte[numBufferChars];

            var result = (int)__get_full_path(relative_path_ascii, resolved_path_ascii);

            if (result != 0)
            {
                Encoding.ASCII.GetChars(resolved_path_ascii, numBufferChars, buffer, numBufferChars);
                return wcslen(buffer);
            }

            return result;
        }

        internal static int GetFileType(SafeFileHandle handle)
        {
            var stdId = handle.DangerousGetHandle().ToInt32();
            if (stdId == STD_OUTPUT_HANDLE || stdId == STD_ERROR_HANDLE)
            {
                return FILE_TYPE_CHAR;
            }

            return FILE_TYPE_DISK;
        }

        // Do not use these directly, use the safe or unsafe versions above.
        // The safe version does not support devices (aka if will only open
        // files on disk), while the unsafe version give you the full semantic
        // of the native version.

        private static SafeFileHandle CreateFile(String lpFileName,
                    int dwDesiredAccess, System.IO.FileShare dwShareMode,
                    SECURITY_ATTRIBUTES securityAttrs, System.IO.FileMode dwCreationDisposition,
                    int dwFlagsAndAttributes, IntPtr hTemplateFile)
        {
            int filed = -1;
            int create_flags = (S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
            int open_flags = 0;
            bool fFileExists = false;

            if (lpFileName == null)
            {
                return new SafeFileHandle(INVALID_HANDLE_VALUE, false);
            }

            unsafe
            {
                fixed (byte* filename_ascii = ToAsciiString(lpFileName))
                {
                    switch ((uint)dwDesiredAccess)
                    {
                        case GENERIC_READ:
                            open_flags |= O_RDONLY;
                            break;
                        case GENERIC_WRITE:
                            open_flags |= O_WRONLY;
                            break;
                        case GENERIC_READ | GENERIC_WRITE:
                            open_flags |= O_RDWR;
                            break;
                        default:
                            return new SafeFileHandle(INVALID_HANDLE_VALUE, false);
                    }

                    switch (dwCreationDisposition)
                    {
                        case System.IO.FileMode.Create:
                            // check whether the file exists
                            if (access(filename_ascii, F_OK) == 0)
                            {
                                fFileExists = true;
                            }

                            open_flags |= O_CREAT | O_TRUNC;
                            break;
                        case System.IO.FileMode.CreateNew:
                            open_flags |= O_CREAT | O_EXCL;
                            break;
                        case System.IO.FileMode.Open:
                            /* don't need to do anything here */
                            break;
                        case System.IO.FileMode.OpenOrCreate:
                            if (access(filename_ascii, F_OK) == 0)
                            {
                                fFileExists = true;
                            }

                            open_flags |= O_CREAT;
                            break;
                        case System.IO.FileMode.Truncate:
                            open_flags |= O_TRUNC;
                            break;
                        default:
                            return new SafeFileHandle(INVALID_HANDLE_VALUE, false);

                    }

                    if ((dwFlagsAndAttributes & FILE_FLAG_NO_BUFFERING) > 0)
                    {
                        open_flags |= O_DIRECT;
                    }

                    open_flags |= O_BINARY;

                    filed = open(filename_ascii, open_flags, (open_flags & O_CREAT) > 0 ? create_flags : 0);
                    if (filed < 0)
                    {
                        return new SafeFileHandle(new IntPtr(0), false);
                    }
                }
            }

#if flock
            var lock_mode = (dwShareMode == 0 /* FILE_SHARE_NONE */) ? LOCK_EX : LOCK_SH;

            if (flock(filed, lock_mode | LOCK_NB) != 0)
            {
                return new SafeFileHandle(INVALID_HANDLE_VALUE, false);
            }
#endif

#if O_DIRECT
            if ((dwFlagsAndAttributes & FILE_FLAG_NO_BUFFERING) > 0)
            {
#if F_NOCACHE
                if (-1 == fcntl(filed, F_NOCACHE, 1))
                {
                    return new SafeFileHandle(INVALID_HANDLE_VALUE, false);
                }
#else
#error Insufficient support for uncached I/O on this platform
#endif
            }
#endif

#if fcntl
            /* make file descriptor close-on-exec; inheritable handles will get
              "uncloseonexeced" in CreateProcess if they are actually being inherited*/
            var ret = fcntl(filed, F_SETFD, __arglist(1));
            if (-1 == ret)
            {
                return new SafeFileHandle(INVALID_HANDLE_VALUE, false);
            }
#endif

            return new SafeFileHandle(new IntPtr(filed), false);
        }

        [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
        internal static bool CloseHandle(IntPtr handle)
        {
            close(handle.ToInt32());
            return true;
        }

        unsafe internal static int ReadFile(SafeFileHandle handle, byte* bytes, int numBytesToRead, out int numBytesRead, IntPtr mustBeZero)
        {
            var r = read(handle.DangerousGetHandle().ToInt32(), bytes, numBytesToRead);
            if (r == -1)
            {
                numBytesRead = 0;
                return 0;
            }

            numBytesRead = r;
            return 1;
        }

        internal static unsafe int WriteFile(SafeFileHandle handle, byte* bytes, int numBytesToWrite, out int numBytesWritten, IntPtr mustBeZero)
        {
            var fd = handle.DangerousGetHandle().ToInt32();
            if (fd == STD_OUTPUT_HANDLE)
            {
                numBytesWritten = write(STDOUT_FILENO, bytes, numBytesToWrite);
                return numBytesWritten < numBytesToWrite ? 0 : 1;
            }
            else if (fd == STD_ERROR_HANDLE)
            {
                numBytesWritten = write(STDERR_FILENO, bytes, numBytesToWrite);
                return numBytesWritten < numBytesToWrite ? 0 : 1;
            }
            else
            {
                var r = write(fd, bytes, numBytesToWrite);
                if (r != -1)
                {
                    numBytesWritten = r;
                    return 1;
                }
            }

            numBytesWritten = 0;
            return 0;
        }

        internal static int GetFileSize(SafeFileHandle hFile, out int highSize)
        {
            highSize = 0;
            unsafe
            {
                var data = new stat_data();
                var returnCode = fstat(hFile.DangerousGetHandle().ToInt32(), &data.st_dev);
                if (returnCode != 0)
                {
                    return 0;
                }

                return data.st_size;
            }
        }

        internal static IntPtr GetStdHandle(int nStdHandle)  // param is NOT a handle, but it returns one!
        {
            return new IntPtr(nStdHandle);
        }

        internal static bool GetFileAttributesEx(String name, int fileInfoLevel, ref WIN32_FILE_ATTRIBUTE_DATA lpFileInformation)
        {
            if (name == null)
            {
                return false;
            }

            unsafe
            {
                var data = new stat_data();
                fixed (byte* filename_ascii = ToAsciiString(name))
                {
                    var returnCode = stat(filename_ascii, &data.st_dev);
                    if (returnCode != 0)
                    {
                        return false;
                    }

                    var fileAttributes = FILE_ATTRIBUTE_NORMAL;

                    // if this is folder, return false
                    if ((data.st_mode & S_IFMT) == S_IFDIR)
                    {
                        fileAttributes = FILE_ATTRIBUTE_DIRECTORY;
                    }

                    lpFileInformation.fileAttributes = (int)fileAttributes;
                    lpFileInformation.fileSizeLow = data.st_size;
                    lpFileInformation.fileSizeHigh = 0;

                    return true;
                }
            }
        }
    }
}

File: Il2Native.Logic/impl/Microsoft/Win32/Win32Native.cpp
#include "CoreLib.h"
#if _MSC_VER
#include <windows.h>
#else
#include <unistd.h>
#include <limits.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <fcntl.h>
#endif

#ifdef GC_PTHREADS
#define FILE_TYPE_CHAR 0x0002
#define FILE_TYPE_DISK 0x0001
#define FILE_TYPE_PIPE 0x0003
#define FILE_TYPE_REMOTE 0x8000
#define FILE_TYPE_UNKNOWN 0x0000
#define GENERIC_READ 0x80000000
#define GENERIC_WRITE 0x40000000
#define FILE_ATTRIBUTE_HIDDEN 0x00000002
#define FILE_ATTRIBUTE_SYSTEM 0x00000004
#define FILE_ATTRIBUTE_DIRECTORY 0x00000010
#define FILE_ATTRIBUTE_ARCHIVE 0x00000020
#define FILE_ATTRIBUTE_DEVICE 0x00000040
#define FILE_ATTRIBUTE_NORMAL 0x00000080
#define FILE_FLAG_NO_BUFFERING 0x20000000
#endif


namespace CoreLib { namespace Microsoft { namespace Win32 { 

	// Method : Microsoft.Win32.Win32Native.SetEvent(Microsoft.Win32.SafeHandles.SafeWaitHandle)
	bool Win32Native::SetEvent(CoreLib::Microsoft::Win32::SafeHandles::SafeWaitHandle* handle)
	{
#ifndef GC_PTHREADS
		return ::SetEvent((HANDLE)handle->handle.ToInt32());
#else
#error NOT IMPLEMENTED YET
#endif
	}

	// Method : Microsoft.Win32.Win32Native.ResetEvent(Microsoft.Win32.SafeHandles.SafeWaitHandle)
	bool Win32Native::ResetEvent(CoreLib::Microsoft::Win32::SafeHandles::SafeWaitHandle* handle)
	{
#ifndef GC_PTHREADS
		return ::ResetEvent((HANDLE)handle->handle.ToInt32());
#else
#error NOT IMPLEMENTED YET
#endif
	}

	// Method : Microsoft.Win32.Win32Native.CreateEvent(Microsoft.Win32.Win32Native.SECURITY_ATTRIBUTES, bool, bool, string)
	CoreLib::Microsoft::Win32::SafeHandles::SafeWaitHandle* Win32Native::CreateEvent(CoreLib::Microsoft::Win32::Win32Native_SECURITY_ATTRIBUTES* lpSecurityAttributes, bool isManualReset, bool initialState, string* name)
	{
#ifndef GC_PTHREADS
		auto hEvent = CreateEventW((LPSECURITY_ATTRIBUTES)lpSecurityAttributes,
			(BOOL)isManualReset,
			(BOOL)initialState,
			(LPCWSTR)(name != nullptr ? &name->m_firstChar : nullptr));

		if (hEvent == (HANDLE)-1)
		{
			return __new<CoreLib::Microsoft::Win32::SafeHandles::SafeWaitHandle>(((CoreLib::System::IntPtr)CoreLib::Microsoft::Win32::Win32Native::INVALID_HANDLE_VALUE), false);
		}

		return __new<CoreLib::Microsoft::Win32::SafeHandles::SafeWaitHandle>(__init<CoreLib::System::IntPtr>(hEvent), false);
#else
#error NOT IMPLEMENTED YET
#endif
	}

	// Method : Microsoft.Win32.Win32Native.GetFullPathName(char*, int, char*, System.IntPtr)
	int32_t Win32Native::GetFullPathName(char16_t* path, int32_t numBufferChars, char16_t* buffer, CoreLib::System::IntPtr mustBeZero)
	{
		if (static_cast<void*>(path) == (void*)nullptr)
		{
			throw __new<CoreLib::System::ArgumentNullException>(u"path"_s, u"path"_s);
		}

#ifndef GC_PTHREADS
		return GetFullPathNameW((LPCWSTR)path, numBufferChars, (LPWSTR)buffer, nullptr);
#elif _WIN32 || _WIN64
		return string::wcslen(_wfullpath(buffer, path, numBufferChars));
#else
		auto path_length = string::wcslen(path);
		auto utf8Enc = CoreLib::System::Text::Encoding::get_UTF8();
		auto byteCount = utf8Enc->GetByteCount(path, path_length);
		auto relative_path_utf8 = reinterpret_cast<uint8_t*>(alloca(byteCount + 1));
		auto bytesReceived = utf8Enc->GetBytes(path, path_length, relative_path_utf8, byteCount);
		auto resolved_path_utf8 = reinterpret_cast<uint8_t*>(alloca(numBufferChars));
		auto result = realpath(reinterpret_cast<const char*>(relative_path_utf8), reinterpret_cast<char*>(resolved_path_utf8));
		if (result != 0)
		{
			utf8Enc->GetChars(resolved_path_utf8, numBufferChars, buffer, numBufferChars);
			return static_cast<int32_t>(string::wcslen(buffer));
		}

		return 0;
#endif
	}

	// Method : Microsoft.Win32.Win32Native.GetStdHandle(int)
	CoreLib::System::IntPtr Win32Native::GetStdHandle(int32_t nStdHandle)
	{
		return __init<CoreLib::System::IntPtr>(nStdHandle);
	}

	// Method : Microsoft.Win32.Win32Native.CreateFile(string, int, System.IO.FileShare, Microsoft.Win32.Win32Native.SECURITY_ATTRIBUTES, System.IO.FileMode, int, System.IntPtr)
	CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle* Win32Native::CreateFile(string* lpFileName, int32_t dwDesiredAccess, CoreLib::System::IO::enum_FileShare dwShareMode, CoreLib::Microsoft::Win32::Win32Native_SECURITY_ATTRIBUTES* securityAttrs, CoreLib::System::IO::enum_FileMode dwCreationDisposition, int32_t dwFlagsAndAttributes, CoreLib::System::IntPtr hTemplateFile)
	{
#ifndef GC_PTHREADS
		auto hFile = CreateFileW((LPCWSTR)&lpFileName->m_firstChar,    // name of the write
			(int32_t)dwDesiredAccess,						 // open for writing
			(int32_t)dwShareMode,							 // do not share
			(LPSECURITY_ATTRIBUTES)securityAttrs,			 // default security
			(int32_t)dwCreationDisposition,					 // create new file only
			(int32_t)dwFlagsAndAttributes,					 // normal file
			(void*)hTemplateFile);							 // no attr. template

		if (hFile == (HANDLE)-1)
		{
			return __new<CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle>(((CoreLib::System::IntPtr)CoreLib::Microsoft::Win32::Win32Native::INVALID_HANDLE_VALUE), false);
		}

		return __new<CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle>(__init<CoreLib::System::IntPtr>(hFile), false);
#else
		int32_t filed = -1;
		int32_t create_flags = (S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
		int32_t open_flags = 0;
		bool fFileExists = false;

		if (lpFileName == nullptr)
		{
			return __new<CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle>(((CoreLib::System::IntPtr)CoreLib::Microsoft::Win32::Win32Native::INVALID_HANDLE_VALUE), false);
		}

		auto path = &lpFileName->m_firstChar;
		auto path_length = string::wcslen(path);
		auto utf8Enc = CoreLib::System::Text::Encoding::get_UTF8();
		auto byteCount = utf8Enc->GetByteCount(path, path_length);
		auto path_urf8 = reinterpret_cast<char*>(alloca(byteCount + 1));
		auto bytesReceived = utf8Enc->GetBytes(path, path_length, (uint8_t*)path_urf8, byteCount);

		switch ((uint32_t)dwDesiredAccess)
		{
		case GENERIC_READ:
			open_flags |= O_RDONLY;
			break;
		case GENERIC_WRITE:
			open_flags |= O_WRONLY;
			break;
		case GENERIC_READ | GENERIC_WRITE:
			open_flags |= O_RDWR;
			break;
		default:
			return __new<CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle>(((CoreLib::System::IntPtr)CoreLib::Microsoft::Win32::Win32Native::INVALID_HANDLE_VALUE), false);
		}

		switch ((CoreLib::System::IO::enum_FileMode)dwCreationDisposition)
		{
		case CoreLib::System::IO::enum_FileMode::c_Create:
			// check whether the file exists
			if (access(path_urf8, F_OK) == 0)
			{
				fFileExists = true;
			}

			open_flags |= O_CREAT | O_TRUNC;
			break;
		case CoreLib::System::IO::enum_FileMode::c_CreateNew:
			open_flags |= O_CREAT | O_EXCL;
			break;
		case CoreLib::System::IO::enum_FileMode::c_Open:
			/* don't need to do anything here */
			break;
		case CoreLib::System::IO::enum_FileMode::c_OpenOrCreate:
			if (access(path_urf8, F_OK) == 0)
			{
				fFileExists = true;
			}

			open_flags |= O_CREAT;
			break;
		case CoreLib::System::IO::enum_FileMode::c_Truncate:
			open_flags |= O_TRUNC;
			break;
		default:
			return __new<CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle>(((CoreLib::System::IntPtr)CoreLib::Microsoft::Win32::Win32Native::INVALID_HANDLE_VALUE), false);
		}

		if ((dwFlagsAndAttributes & FILE_FLAG_NO_BUFFERING) > 0)
		{
			open_flags |= O_DIRECT;
		}

		filed = open(path_urf8, open_flags, (open_flags & O_CREAT) > 0 ? create_flags : 0);
		if (filed < 0)
		{
			return __new<CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle>(__init<CoreLib::System::IntPtr>(0), false);
		}

#if flock
		auto lock_mode = (dwShareMode == 0 /* FILE_SHARE_NONE */) ? LOCK_EX : LOCK_SH;
		if (flock(filed, lock_mode | LOCK_NB) != 0)
		{
			return __new<CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle>(INVALID_HANDLE_VALUE, false);
		}
#endif

#if O_DIRECT
		if ((dwFlagsAndAttributes & FILE_FLAG_NO_BUFFERING) > 0)
		{
#if F_NOCACHE
			if (-1 == fcntl(filed, F_NOCACHE, 1))
			{
				return __new<CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle>(INVALID_HANDLE_VALUE, false);
			}
#else
			////#error Insufficient support for uncached I/O on this platform
#endif
		}
#endif

#if fcntl
		/* make file descriptor close-on-exec; inheritable handles will get
		"uncloseonexeced" in CreateProcess if they are actually being inherited*/
		auto ret = fcntl(filed, F_SETFD, 1);
		if (-1 == ret)
		{
			return __new<CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle>(INVALID_HANDLE_VALUE, false);
		}
#endif

		return __new<CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle>(__init<CoreLib::System::IntPtr>(filed), false);
#endif
	}

	// Method : Microsoft.Win32.Win32Native.CloseHandle(System.IntPtr)
	bool Win32Native::CloseHandle(CoreLib::System::IntPtr handle)
	{
#ifndef GC_PTHREADS
		return ::CloseHandle((HANDLE)handle.ToInt32());
#else
		// TODO: finish it for Event, write wrapper with Destroy function
		close(handle.ToInt32());
		return true;
#endif
	}

	// Method : Microsoft.Win32.Win32Native.GetFileType(Microsoft.Win32.SafeHandles.SafeFileHandle)
	int32_t Win32Native::GetFileType(CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle* handle)
	{
		auto stdId = handle->DangerousGetHandle()->ToInt32();
		if (stdId == -11 || stdId == -12)
		{
			return FILE_TYPE_CHAR;
		}

		return FILE_TYPE_DISK;
	}

	// Method : Microsoft.Win32.Win32Native.GetFileSize(Microsoft.Win32.SafeHandles.SafeFileHandle, out int)
	int32_t Win32Native::GetFileSize_Out(CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle* hFile, int32_t& highSize)
	{
#ifndef GC_PTHREADS
		return GetFileSize((HANDLE)hFile->DangerousGetHandle()->ToInt32(), (LPDWORD)highSize);
#else
		highSize = 0;
		struct stat data;
		auto returnCode = fstat(hFile->DangerousGetHandle()->ToInt32(), &data);
		if (returnCode != 0)
		{
			return 0;
		}

		return data.st_size;
#endif
	}

	// Method : Microsoft.Win32.Win32Native.ReadFile(Microsoft.Win32.SafeHandles.SafeFileHandle, byte*, int, out int, System.IntPtr)
	int32_t Win32Native::ReadFile_Out(CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle* handle, uint8_t* bytes, int32_t numBytesToRead, int32_t& numBytesRead, CoreLib::System::IntPtr mustBeZero)
	{
		auto fd = handle->DangerousGetHandle()->ToInt32();
#if _MSC_VER
		return (int32_t) ::ReadFile((HANDLE)fd, (LPVOID)bytes, numBytesToRead, (LPDWORD)&numBytesRead, nullptr);
#else
		auto r = read(fd, bytes, numBytesToRead);
		if (r == -1)
		{
			numBytesRead = 0;
			return 0;
		}

		numBytesRead = r;
		return 1;
#endif
	}

	// Method : Microsoft.Win32.Win32Native.WriteFile(Microsoft.Win32.SafeHandles.SafeFileHandle, byte*, int, out int, System.IntPtr)
	int32_t Win32Native::WriteFile_Out(CoreLib::Microsoft::Win32::SafeHandles::SafeFileHandle* handle, uint8_t* bytes, int32_t numBytesToWrite, int32_t& numBytesWritten, CoreLib::System::IntPtr mustBeZero)
	{
		auto fd = handle->DangerousGetHandle()->ToInt32();
#ifndef GC_PTHREADS
		return (int32_t) ::WriteFile((HANDLE)fd, (LPCVOID)bytes, numBytesToWrite, (LPDWORD)&numBytesWritten, nullptr);
#else
		if (fd == -11)
		{
			numBytesWritten = write(STDOUT_FILENO, bytes, numBytesToWrite);
			return numBytesWritten < numBytesToWrite ? 0 : 1;
		}
		else if (fd == -12)
		{
			numBytesWritten = write(STDERR_FILENO, bytes, numBytesToWrite);
			return numBytesWritten < numBytesToWrite ? 0 : 1;
		}
		else
		{
			auto r = write(fd, bytes, numBytesToWrite);
			if (r != -1)
			{
				numBytesWritten = r;
				return 1;
			}
		}

		numBytesWritten = 0;
		return 0;
#endif
	}

	// Method : Microsoft.Win32.Win32Native.CreateDirectory(string, Microsoft.Win32.Win32Native.SECURITY_ATTRIBUTES)
	bool Win32Native::CreateDirectory(string* path, CoreLib::Microsoft::Win32::Win32Native_SECURITY_ATTRIBUTES* lpSecurityAttributes)
	{
		throw 3221274624U;
	}

	// Method : Microsoft.Win32.Win32Native.GetCurrentDirectory(int, System.Text.StringBuilder)
	int32_t Win32Native::GetCurrentDirectory(int32_t nBufferLength, CoreLib::System::Text::StringBuilder* lpBuffer)
	{
#ifndef GC_PTHREADS
		auto buffer = reinterpret_cast<char16_t*>(alloca(nBufferLength + 1));
		auto result = ::GetCurrentDirectoryW(nBufferLength, (LPWSTR)buffer);
		if (result != 0)
		{
			lpBuffer->Append(__wchar_t_to_string(buffer));
		}

		return result;
#else
		auto dwDirLen = 0;
		auto dwLastError = 0;

		char  *current_dir;

		/* NULL first arg means getcwd will allocate the string */
		current_dir = getcwd( NULL, PATH_MAX + 1 );

		if ( !current_dir )
		{
			// TODO: finish it
			////dwLastError = ERROR_INTERNAL_ERROR;
			goto done;
		}

		dwDirLen = strlen( current_dir );

		/* if the supplied buffer isn't long enough, return the required
		length, including room for the NULL terminator */
		if ( nBufferLength <= dwDirLen )
		{
			++dwDirLen; /* include space for the NULL */
			goto done;
		}
		else
		{
			lpBuffer->Append(__utf8_to_string(current_dir));
		}

done:
		free( current_dir );

		if ( dwLastError )
		{
			// TODO: finish it
			////SetLastError(dwLastError);
		}

		return dwDirLen;
#endif
	}

	// Method : Microsoft.Win32.Win32Native.GetFileAttributesEx(string, int, ref Microsoft.Win32.Win32Native.WIN32_FILE_ATTRIBUTE_DATA)
	bool Win32Native::GetFileAttributesEx_Ref(string* name, int32_t fileInfoLevel, CoreLib::Microsoft::Win32::Win32Native_WIN32_FILE_ATTRIBUTE_DATA& lpFileInformation)
	{
#ifndef GC_PTHREADS
		return ::GetFileAttributesExW((LPCWSTR)&name->m_firstChar, (GET_FILEEX_INFO_LEVELS)fileInfoLevel, &lpFileInformation);
#else
		auto filename = &name->m_firstChar;
		auto filename_length = string::wcslen(filename);
		auto utf8Enc = CoreLib::System::Text::Encoding::get_UTF8();
		auto byteCount = utf8Enc->GetByteCount(filename, filename_length);
		auto filename_urf8 = reinterpret_cast<char*>(alloca(byteCount + 1));
		auto bytesReceived = utf8Enc->GetBytes(filename, filename_length, (uint8_t*)filename_urf8, byteCount);

		struct stat data;
		auto return_code = stat(filename_urf8, &data);
		if (return_code != 0)
		{
			return false;
		}

		auto fileAttributes = FILE_ATTRIBUTE_NORMAL;

		// if this is folder, return false
		if ((data.st_mode & S_IFMT) == S_IFDIR)
		{
			fileAttributes = FILE_ATTRIBUTE_DIRECTORY;
		}

		lpFileInformation.fileAttributes = (int)fileAttributes;
		lpFileInformation.fileSizeLow = data.st_size;
		lpFileInformation.fileSizeHigh = 0;

		return true;
#endif
	}

	// Method : Microsoft.Win32.Win32Native.RemoveDirectory(string)
	bool Win32Native::RemoveDirectory(string* path)
	{
		throw 3221274624U;
	}

	// Method : Microsoft.Win32.Win32Native.SetCurrentDirectory(string)
	bool Win32Native::SetCurrentDirectory(string* path)
	{
#ifndef GC_PTHREADS
		return ::SetCurrentDirectoryW((LPCWSTR)&path->m_firstChar) > 0;
#else
		auto path_ptr = &path->m_firstChar;
		auto path_length = string::wcslen(path_ptr);
		auto utf8Enc = CoreLib::System::Text::Encoding::get_UTF8();
		auto byteCount = utf8Enc->GetByteCount(path_ptr, path_length);
		auto path_urf8 = reinterpret_cast<char*>(alloca(byteCount + 1));
		auto bytesReceived = utf8Enc->GetBytes(path_ptr, path_length, (uint8_t*)path_urf8, byteCount);

		return chdir(path_urf8) == 0;
#endif
	}
}}}
